ouroboros-network/src/Data/Sequence/Strict.hs

```diff
@@ -0,0 +1,221 @@
+-- sequence.
+pattern (:<|) :: a -> StrictSeq a -> StrictSeq a
+pattern x :<| xs <- (viewFront -> Just (x, xs))
+  where
+    x :<| xs = x <| xs
+
+-- | A bidirectional pattern synonym viewing the rear of a non-empty
+-- sequence.
+pattern (:|>) :: StrictSeq a -> a -> StrictSeq a
+pattern xs :|> x <- (viewBack -> Just (xs, x))
+  where
+    xs :|> x = xs |> x
+
+{-------------------------------------------------------------------------------
+  Construction
+-------------------------------------------------------------------------------}
+
+-- | \( O(1) \). The empty sequence.
+empty :: StrictSeq a
+empty = StrictSeq (Seq.empty)
+
+-- | \( O(1) \). A singleton sequence.
+singleton :: a -> StrictSeq a
+singleton !x =  StrictSeq (Seq.singleton x)
+
+-- | \( O(1) \). Add an element to the left end of a sequence.
+-- Mnemonic: a triangle with the single element at the pointy end.
+(<|) :: a -> StrictSeq a -> StrictSeq a
+(!x) <| StrictSeq s = StrictSeq (x Seq.<| s)
+
+-- | \( O(1) \). Add an element to the right end of a sequence.
+-- Mnemonic: a triangle with the single element at the pointy end.
+(|>) :: StrictSeq a -> a -> StrictSeq a
+StrictSeq s |> (!x) = StrictSeq (s Seq.|> x)
+
+-- | \( O(\log(\min(n_1,n_2))) \). Concatenate two sequences.
+--
+-- @intricate: This probably isn't right.
+(><) :: StrictSeq a -> StrictSeq a -> StrictSeq a
+StrictSeq xs >< StrictSeq ys = StrictSeq (xs Seq.>< ys)
+
+fromList :: [a] -> StrictSeq a
+fromList !xs = foldl' (|>) (StrictSeq Seq.Empty) xs
+
+{-------------------------------------------------------------------------------
+  Deconstruction
+-------------------------------------------------------------------------------}
+
+-- | \( O(1) \). Is this the empty sequence?
+null :: StrictSeq a -> Bool
+null (StrictSeq xs) = Seq.null xs
+
+-- | \( O(1) \). The number of elements in the sequence.
+length :: StrictSeq a -> Int
+length (StrictSeq xs) = Seq.length xs
+
+{-------------------------------------------------------------------------------
+  Scans
+-------------------------------------------------------------------------------}
+
+-- | 'scanl' is similar to 'foldl', but returns a sequence of reduced
+-- values from the left:
+--
+-- > scanl f z (fromList [x1, x2, ...]) = fromList [z, z `f` x1, (z `f` x1) `f` x2, ...]
+scanl :: (a -> b -> a) -> a -> StrictSeq b -> StrictSeq a
+scanl f !z0 (StrictSeq xs) = StrictSeq $ forceElemsToWHNF $ Seq.scanl f z0 xs
+-- scanl f z0 xs = z0 <| snd (mapAccumL (\ x z -> let x' = f x z in (x', x')) z0 xs)
+
+{-------------------------------------------------------------------------------
+  Sublists
+-------------------------------------------------------------------------------}
+
+-- | \( O(i) \) where \( i \) is the prefix length.  @'dropWhileL' p xs@ returns
+-- the suffix remaining after @'takeWhileL' p xs@.
+dropWhileL :: (a -> Bool) -> StrictSeq a -> StrictSeq a
+dropWhileL p (StrictSeq xs) = StrictSeq (Seq.dropWhileL p xs)
```

Comment:
BTW, if you're just forwarding to `Seq` without any forcing, you could also use `coerce`, e.g.
```haskell
dropWhileL :: (a -> Bool) -> StrictSeq a -> StrictSeq a
dropWhileL = coerce Seq.dropWhileL
```

Reply:
I didn't know that, thanks.